monitoring/monitoring/monitoring.py

```diff
@@ -0,0 +1,219 @@
+
+    try:
+        time_period = datetime.datetime.strptime(time_period_query, date_format)
+        month = time_period.month
+        year = time_period.year
+        if time_period < datetime.datetime.strptime('07/2020', date_format):
+            return await parse_error(f"Invalid value for time_period '{time_period_query}'; time periods before 07/2020 are not supported.")
+    except ValueError:
+        return await parse_error(f"Invalid value for time_period '{time_period_query}'; must be in the format of MM/YYYY.")
+
+    db = app['db']
+    records = db.execute_and_fetchall('SELECT * FROM monitoring_billing_data WHERE year = %s AND month = %s;', (year, month))
+    records = [record async for record in records]
+
+    cost_by_service, compute_cost_breakdown, cost_by_sku_label = format_data(records)
+
+    return (cost_by_service, compute_cost_breakdown, cost_by_sku_label, time_period_query)
+
+
+@routes.get('/billing')
+@web_authenticated_developers_only()
+async def billing(request: web.Request, userdata) -> web.Response:  # pylint: disable=unused-argument
+    cost_by_service, compute_cost_breakdown, cost_by_sku_label, time_period_query = await _billing(request)
+    context = {
+        'cost_by_service': cost_by_service,
+        'compute_cost_breakdown': compute_cost_breakdown,
+        'cost_by_sku_label': cost_by_sku_label,
+        'time_period': time_period_query
+    }
+    return await render_template('monitoring', request, userdata, 'billing.html', context)
+
+
+async def query_billing_body(app):
+    db = app['db']
+    bigquery_client = app['bigquery_client']
+
+    async def _query(dt):
+        month = dt.month
+        year = dt.year
+
+        start = datetime.date(year, month, 1)
+        _, last_day_of_month = calendar.monthrange(year, month)
+        end = datetime.date(year, month, last_day_of_month) + datetime.timedelta(days=7)
+
+        date_format = '%Y-%m-%d'
+        start_str = datetime.date.strftime(start, date_format)
+        end_str = datetime.date.strftime(end, date_format)
+
+        invoice_month = datetime.date.strftime(start, '%Y%m')
+
+        cmd = f'''
+SELECT service.description as service_description, sku.description as sku_description, SUM(cost) as cost,
+CASE
+  WHEN EXISTS(SELECT 1 FROM UNNEST(labels) WHERE key = "namespace" and value = "default") THEN "batch-production"
+  WHEN EXISTS(SELECT 1 FROM UNNEST(labels) WHERE key = "namespace" and value LIKE '%pr-%') THEN "batch-test"
+  WHEN EXISTS(SELECT 1 FROM UNNEST(labels) WHERE key = "namespace") THEN "batch-dev"
+  WHEN EXISTS(SELECT 1 FROM UNNEST(labels) WHERE key = "role" and value LIKE 'vdc') THEN "k8s"
+  WHEN service.description = "Compute Engine" THEN "unknown"
+  ELSE NULL
+END AS source
+FROM `broad-ctsa.hail_billing.gcp_billing_export_v1_0055E5_9CA197_B9B894`
+WHERE DATE(_PARTITIONTIME) >= "{start_str}" AND DATE(_PARTITIONTIME) <= "{end_str}" AND project.name = "hail-vdc" AND invoice.month = "{invoice_month}"
```

Comment:
Can you tell me more about what _PARTITIONTIME is and why we need to use it here? Seems like invoice.month would be sufficient, but I assume not.

Reply:
I think it's not sufficient and we'll scan all the data. I did a test query plan for July and without the partition time it will scan 128 MB instead of 32 MB. Some examples have the partition time and others don't. https://cloud.google.com/billing/docs/how-to/bq-examples#query-by-project-by-invoice

https://cloud.google.com/bigquery/docs/querying-partitioned-tables#ingestion-time_partitioned_table_pseudo_columns